atef/walk.py

```diff
@@ -0,0 +1,175 @@
+"""
+Helpers for walking dataclasses.
+
+If relying on Prepared dataclass walk methods (walk_groups, walk_comparisons),
+note that if a configuration fails to prepare, they will be skipped
+"""
+from __future__ import annotations
+
+from typing import Any, Generator, List, Tuple, Union
+
+from atef.check import Comparison
+from atef.config import (Configuration, PreparedComparison,
+                         PreparedConfiguration, PreparedFile)
+from atef.procedure import (PreparedProcedureFile, PreparedProcedureStep,
+                            ProcedureStep)
+
+
+def walk_config_file(
+    config: Union[PreparedFile, PreparedConfiguration, PreparedComparison],
+    level: int = 0
+) -> Generator[Tuple[Any, int], None, None]:
+    """
+    Yields each config and comparison and its depth
+    Performs a recursive depth-first search
+
+    Parameters
+    ----------
+    config : Union[PreparedFile, PreparedConfiguration, PreparedComparison]
+        the configuration or comparison to walk
+    level : int, optional
+        the current recursion depth, by default 0
+
+    Yields
+    ------
+    Generator[Tuple[Any, int], None, None]
+    """
+    yield config, level
+    if isinstance(config, PreparedFile):
+        yield from walk_config_file(config.root, level=level+1)
+    elif isinstance(config, PreparedConfiguration):
+        if hasattr(config, 'configs'):
+            for conf in config.configs:
+                yield from walk_config_file(conf, level=level+1)
+        if hasattr(config, 'comparisons'):
+            for comp in config.comparisons:
+                yield from walk_config_file(comp, level=level+1)
+
+
+def walk_procedure_file(
+    config: Union[PreparedProcedureFile, PreparedProcedureStep, PreparedComparison],
+    level: int = 0
+) -> Generator[Tuple[Any, int], None, None]:
```

Comment:
Careful with the `Any` annotation - it specifically says "don't bother checking the type for this" and it's actually a bit worse than some alternatives ([ref](https://stackoverflow.com/questions/39817081/typing-any-vs-object))

Might be good to make unions for possible things to come out of this.